Supporting classifiers and packaging in ArtifactOrProject

roblg opened the pull request:
Sorry for the churn on this. This is based on my comments in https://github.com/johnynek/bazel-deps/issues/69#issuecomment-372370151, and works on our codebase, and should work for multiple classifers. It supports the "front half" of specifying packaging and classifiers. (Classifiers should "just work", but there's no work done yet to distinguish between `jar` packaging and non-`jar` packaging in the BUILD files, so if you depend on a non-jar package, I believe the BUILD files will be broken.

* this works on our codebase. The only change is the addition of packaging+classifier
  in target names and bind
* needs some real tests
* needs consensus on the palatability of this .yaml syntax
* I probably broke scala somehow
* I snuck in support for `exclude` of particular packaging/classifiers

Definitely needs work:
* non-jar `packaging` is added to .bzl and BUILD files, but results in broken BUILD files
* needs test for the changes around excludes
* I haven't tested .yaml _generation_ yet, or the maven Tool.scala

roblg commented on src/scala/com/github/johnynek/bazel_deps/Resolver.scala:
```diff
@@ -63,7 +63,11 @@ class Resolver(servers: List[MavenServer], resolverCachePath: Path) {
+        elem.group.asString,
+        elem.artifact.getArtifact,
+        elem.artifact.getClassifier.orNull,
+        elem.artifact.getPackaging)
```
Example of exclusions including `packaging`:

``
  com.amazonaws:
    DynamoDBLocal:
      lang: java
      version: "1.11.86"
      exclude:
        - "com.almworks.sqlite4java:sqlite4java-win32-x86:dll"
        - "com.almworks.sqlite4java:sqlite4java-win32-x64:dll"
        - "com.almworks.sqlite4java:libsqlite4java-osx:dylib"
        - "com.almworks.sqlite4java:libsqlite4java-linux-i386:so"
        - "com.almworks.sqlite4java:libsqlite4java-linux-amd64:so"
```

I suspect, but haven't tried that you could also exclude `"com.almworks.sqlite4java:libsqlite4java-linux-i386:*"` (asterisk at the end) to exclude _all_ packaging types for that dependency (if more than one were present). Seems to be a feature of aether.

kevingessner commented on src/scala/com/github/johnynek/bazel_deps/DepsModel.scala:
```diff
@@ -159,12 +172,20 @@ case class ArtifactOrProject(asString: String) {
+    str.split(":") match {
+      case Array(a, p, c) => ArtifactOrProject(a, p, Some(c))
+      case Array(a, p) => ArtifactOrProject(a, p, None)
+      case Array(a) => ArtifactOrProject(a, "jar", None)
```
I see `"jar"` repeated many times as the default packaging.  Replace them with a `val defaultPackaging = "jar"` for readability?

kevingessner commented:
I'm digging this PR.  I tried the branch to import a few scala jars with classifiers (e.g. `com.twitter:inject-core_2.11:jar:tests`) and it works great.  Let's get it merged!

roblg commented:
I added some more tests, and made some tweaks to the PR (functionality is the same)

I'm planning to look at Tool.scala next.

This is still a problem:

> * non-jar `packaging` is added to .bzl and BUILD files, but results in broken BUILD files

I'm tempted to _not_ address that in this PR, partly because I'm not sure what the appropriate course of action is (e.g., is it sufficient to just add anything with a non-jar packaging to `runtime_deps` instead of `deps`? Does that even work? The one place we use non-jar deps, we appear to add them as `data` for a test target). I'd be happy to explicitly disallow non-jar packaging in the dependencies.yaml file (except for in `excludes`). It seems like the appropriate place to do that would be in the decoding phase, but I couldn't figure out how that would work. I can always pop in a hack after everything is decoded if we think it'll only be short-lived too.

Sounds like @kevingessner is on board w/ the syntax. @johnynek what're your thoughts?

jin commented:
Just tried this out, it works for `aar` on maven.google.com too.

I did run into an issue where the `aar` transitive dependencies of an `aar` resolves incorrectly to `jar`. The debugger led me to Aether's resolving mechanism, so I don't think this PR has anything to do with that.

jin commented:
Dug a little bit more and found an issue that might affect this PR.

Even if the correct packaging type is passed to Aether, if the transitive dependencies of the artifact do not specify the <type /> in the <dependencies> section in the `pom.xml`, Aether defaults the type to `jar`. This causes errors like:

```
failed to find sha of com.android.support.test:rules:aar:1.0.1: org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact com.android.support.test:runner:jar:1.0.1 in google (https://maven.google.com)
Exception in thread "main" org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact com.android.support.test:runner:jar:1.0.1 in google (https://maven.google.com)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:444)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:246)
        at org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifacts(DefaultRepositorySystem.java:302)
        at com.github.johnynek.bazel_deps.Resolver$$anonfun$getExt$1$1.apply(Resolver.scala:105)
        at com.github.johnynek.bazel_deps.Resolver$$anonfun$getExt$1$1.apply(Resolver.scala:103)
        at scala.util.Try$.apply(Try.scala:192)
        at com.github.johnynek.bazel_deps.Resolver.getExt$1(Resolver.scala:103)
        at com.github.johnynek.bazel_deps.Resolver.getShas(Resolver.scala:117)
        at com.github.johnynek.bazel_deps.MakeDeps$.apply(MakeDeps.scala:88)
        at com.github.johnynek.bazel_deps.ParseProject$.main(ParseProject.scala:11)
        at com.github.johnynek.bazel_deps.ParseProject.main(ParseProject.scala)
Caused by: org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact com.android.support.test:runner:jar:1.0.1 in google (https://maven.google.com)
        at org.eclipse.aether.connector.basic.ArtifactTransportListener.transferFailed(ArtifactTransportListener.java:39)
        at org.eclipse.aether.connector.basic.BasicRepositoryConnector$TaskRunner.run(BasicRepositoryConnector.java:355)
        at org.eclipse.aether.util.concurrency.RunnableErrorForwarder$1.run(RunnableErrorForwarder.java:67)
        at org.eclipse.aether.connector.basic.BasicRepositoryConnector$DirectExecutor.execute(BasicRepositoryConnector.java:581)
        at org.eclipse.aether.connector.basic.BasicRepositoryConnector.get(BasicRepositoryConnector.java:249)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.performDownloads(DefaultArtifactResolver.java:520)
        at org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:421)
```

roblg commented:
@jin do you know if there's something that can be done in this PR to address that? Or is it something upstream that we'll just have to live with for now?

johnynek commented:
note, I found some issues with aether sometimes not finding all dependencies. We implemented a coursier backend #145 .

It  may do better for this. I would like to get this issue solved, sorry I've been so quiet about this, I don't personally use the feature so I have less intuition about the corner cases.

jin commented:
> Or is it something upstream that we'll just have to live with for now?

I think this has to be an upstream fix, since it's Aether doing the pom parsing and making the assumption that untyped artifacts are jars. The other upstream fix is to add the types directly to Google Maven poms.

johnynek commented:
@roblg in the short term, it would be nice to support classifiers on both backends, but it may be that we will abandon aether. It seems it is no longer developed (was sent to apache and renamed). So, not sure the best way forward.

roblg commented:
Updated the PR to reflect the Resolver->AetherResolver move, and added classifier support to CoursierResolver.

Coursier exposed several fun data bugs in the public maven repository, and in our internal maven repository. :) No bugs in coursier itself though (afaict)! Lots more runtime_deps magically appeared on the switchover.

johnynek commented on src/scala/com/github/johnynek/bazel_deps/DepsModel.scala:
```diff
@@ -312,16 +327,48 @@ object Version {
+      case Array(a, p, c) => MavenArtifactId(a, p, Some(c))
+      case Array(a, p) => MavenArtifactId(a, p, None)
+      case Array(a) => MavenArtifactId(a, defaultPackaging, None)
+    }
```
can we add a better exception message here in the `case _ => `?

johnynek commented on src/scala/com/github/johnynek/bazel_deps/maven/Tool.scala:
```diff
@@ -81,16 +83,22 @@ object Tool {
           props.getOrElse(k, "${" + k + "}")
         })
       }
+      def ropt(s: Option[String]): Option[String] = s match {
```
isn't this just `s.map(r)`? can we use that instead?

roblg replied:
Doh. Yes, thanks.

johnynek commented on test/scala/com/github/johnynek/bazel_deps/ModelTest.scala:
```diff
@@ -64,4 +64,44 @@ class ModelTest extends FunSuite {
+    assert(MavenArtifactId("foo:jar").addSuffix("_1").asString == "foo_1")
+  }
+
+  test("MavenArtifactId packaging and classifier defaults") {
```
can you add a comment as to what kinds of tests you want to put here?

roblg replied:
I started to write this test, and then realized that the cases I wanted to cover are already covered by the test on line 68 (`packaging and classifier are extracted properly parsed in MavenArtifactId`), and I just forgot to remove it. Thanks!

johnynek commented:
can we update the README with an example of how to use this feature?

Also, note I have a coursier resolver test currently, can you add a test using a classifier?

roblg commented on test/scala/com/github/johnynek/bazel_deps/CoursierTest.scala:
```diff
@@ -38,4 +38,59 @@ dependencies:
+    assert(graph.nodes.contains(
+      MavenCoordinate(
+        MavenGroup("net.sf.json-lib"), MavenArtifactId("json-lib", "jar", "jdk15"), Version("2.4"))))
+  }
```
@johnynek does testing `graph.nodes.contains` here seem sufficient, or is there something more thorough that would be betteR?

johnynek replied:
This seems great. Can we add a test for exclusion of a non-jar resource (which we advertise support for) and then I'll merge.

roblg replied:
Done!

johnynek commented:
Thank you for staying with this!

roblg commented:
No problem! I was going to be using this branch internally either way. Makes my life much easier if it's on master. ;) Thanks for putting up with my (barely) rudimentary Scala!
